### DISCUSSION TITLE
Cannot find module 'inversify' or its corresponding type declaration

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1230] [Q&A] [answered]
During my migration to GLSP 2.1.0 I run into some problem with the VSCode Integration.

At least I can't get rid of the error message

```
open-bpmn-vscode-webview: $ tsc -b && yarn bundle
open-bpmn-vscode-webview: src/app.ts(20,27): error TS2307: Cannot find module 'inversify' or its corresponding type declarations.
error Command failed with exit code 1.
```

And I can't figure out what the reason for this error could be. At least I assume that I have to add a missing dependency. In my root package.json file I have

```
  "devDependencies": {
    "@eclipse-glsp/dev": "2.0.0",
    "@types/node": "16",
    "@types/vscode": "^1.54.0",
    "lerna": "^6.6.2",
    "typescript": "^5.2.2"
  },
```

Does this makes sense? I did not know what ` "@eclipse-glsp/dev": "2.0.0",` is good for. 

In my webview module - which is causing the problem I use the following dependencies:

```
  "devDependencies": {
    "@eclipse-glsp/vscode-integration-webview": "2.1.0-next.74",
    "@open-bpmn/open-bpmn-glsp": "1.2.0",
    "@vscode/codicons": "^0.0.25",
    "circular-dependency-plugin": "^5.2.2",
    "css-loader": "^6.7.1",
    "ignore-loader": "^0.1.2",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "reflect-metadata": "^0.1.13",
    "source-map-loader": "^4.0.1",
    "style-loader": "^3.3.1",
    "ts-loader": "^9.4.2",
    "webpack": "^5.75.0",
    "webpack-cli": "^5.0.1",
    "webpack-merge": "^5.9.0"
  }
```

This is all aligned to the examples:

 - [Minimal task example integration for VS Code](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/node-json-vscode/tasklist-vscode)
- [Example VS Code extension for the workflow diagram](https://github.com/eclipse-glsp/glsp-vscode-integration/tree/master/example/workflow)

...hm... I have no more ideas :-(

**Comment by tortmayr** [comment]
Hi @rsoika,

with 2.x GLSP/Sprotty no longer depended on a fixed version of inversify. Instead a compatible range is declared as `peerDependency`.
It is than up to t he adopter to provide a matching version. This ensures that exactly one version of inversify is used in the entire project which is crucial because otherwise the DI configuration might fail.

In your case you simple have to a add a matching version as part of the "devDependencies" e.g. "6.0.1".

You can also checkout the first iteration of our [Migration Guide](https://github.com/eclipse-glsp/glsp/blob/master/MIGRATION_GUIDE.md). It currently covers the migration steps for the client and the Theia integration and we are currently working on extending the guide for the servers, VS Code and Eclipse.

Regarding the `@eclipse-glsp/dev` package I will answer that question in https://github.com/eclipse-glsp/glsp/discussions/1231.

**Comment by rsoika** [comment]
Thanks for your answer. Now I understand my problem much better. I did not recognize that the [Migration Guide](https://github.com/eclipse-glsp/glsp/blob/master/MIGRATION_GUIDE.md) has involved so much.

**Comment by rsoika** [comment]
Hi @tortmayr , I still have the problem with the in `inversify` package or maybe with the package `vscode-jsonrpc`.

I try (now for several weeks) to get my GLSP 2.1.x Project integrated into VS-Code. Adding the `@eclipse-glsp/dev devDependecy` solves the compiler problem as we discussed [here](https://github.com/eclipse-glsp/glsp/discussions/1231). I also joined [this discussion](https://github.com/eclipse-glsp/glsp/discussions/1224) and try to write a kind of [vscode-integration-guide](https://github.com/imixs/open-bpmn-vscode-integration/blob/main/QUICKSTARTGUIDE.md) . But my extension still does not work.

When I try to load a diagram I get in the VS-Code client the following error message:


```
TypeError: Reflect.hasOwnMetadata is not a function
	at /home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:15946:21
	at __webpack_modules__.../node_modules/@eclipse-glsp/protocol/lib/client-server-protocol/jsonrpc/base-jsonrpc-glsp-client.js.__decorate (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:2049:95)
	at ../node_modules/@eclipse-glsp/protocol/lib/client-server-protocol/jsonrpc/base-jsonrpc-glsp-client.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:2215:51)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at ../node_modules/@eclipse-glsp/protocol/lib/client-server-protocol/jsonrpc/index.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:2352:14)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at ../node_modules/@eclipse-glsp/protocol/lib/client-server-protocol/index.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:2032:14)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at ../node_modules/@eclipse-glsp/protocol/lib/index.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:2690:14)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at ../node_modules/@eclipse-glsp/vscode-integration/lib/common/index.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:4624:14)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at ../node_modules/@eclipse-glsp/vscode-integration/lib/node/index.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:5095:14)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at ../node_modules/@eclipse-glsp/vscode-integration/node.js (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:5502:18)
	at __webpack_require__ (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33345:42)
	at /home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33419:16
	at /home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33482:3
	at Object.<anonymous> (/home/rsoika/git/open-bpmn-vscode-integration/extension/dist/open-bpmn-extension.js:33485:12)
	at Module._compile (node:internal/modules/cjs/loader:1316:14)
	at Module._extensions..js (node:internal/modules/cjs/loader:1371:10)
	at Module.load (node:internal/modules/cjs/loader:1171:32)
	at Module._load (node:internal/modules/cjs/loader:1012:12)
	at f._load (node:electron/js2c/asar_bundle:2:13377)
	at c._load (/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:171:5632)
	at n._load (/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:168:29551)
	at t._load (/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:133:35226)
	at Module.require (node:internal/modules/cjs/loader:1195:19)
	at require (node:internal/modules/cjs/helpers:110:18)
	at Function.i [as __$__nodeRequire] (/usr/share/code/resources/app/out/vs/loader.js:5:98)
	at f.vb (/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:168:30730)
	at async Promise.all (index 0)
```

So this looks to me that the extension is still missing the packages. But I can't figure out what I am doing wrong. I think now it should be the `webpack.config.js` in my /extension module. This configuration file looks currently like this:


```javascript
'use strict';

const path = require('path');
const webpack = require('webpack');
const CopyPlugin = require('copy-webpack-plugin');
const nodeModules = path.resolve(__dirname, '../node_modules');

/**@type {import('webpack').Configuration}*/
const config = {
    target: 'node',

    entry: path.resolve(__dirname, 'src/open-bpmn-extension.ts'),
    output: {
        path: path.resolve(__dirname, 'dist'),
        filename: 'open-bpmn-extension.js',
        libraryTarget: 'commonjs2'
    },
    devtool: 'source-map',
    externals: {
        vscode: 'commonjs vscode'
    },
    mode: 'development',
    resolve: {
        extensions: ['.ts', '.js']
    },
    module: {
        rules: [
            {
                test: /\.ts$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'ts-loader'
                    }
                ]
            }
        ]
    },
    plugins: [
        new CopyPlugin({
            patterns: [                             
                {
                    from: path.resolve(__dirname, '..', 'webview', 'dist')
                }
            ]
        })
    ],
    ignoreWarnings: [/Can't resolve .* in '.*ws\/lib'/],
    performance: {
        hints: false
    }
};
module.exports = config;
```

The packages `inversify` and `vscode-jsonrpc` are all part of my `../node_modules` directory. But maybe these packages did not become part of the final webpack file?  Could this be the reason for the errors?

**Comment by tortmayr** [comment]
Inversify is only used in the webview (i.e. client) part of the vscode integration. You posted the webpack config of your extension.
You should have another config in your "webview" package.
This error in particular can occur if you did not correctly import "reflect-metadata" in the entry point class of you webview component.
For example, have a look at the `glsp-vcode-integration` template: https://github.com/eclipse-glsp/glsp-examples/blob/335106e15aa0924ef900864351ce471c18fdd672/project-templates/node-json-vscode/tasklist-vscode/webview/src/app.ts#L17

The first import in the "app" class is "reflect-metadata" which is needed for inversify to work properly.

**Reply by rsoika** [reply]
Hm..., I followed the [glsp-vscode-integration/workflow example](https://github.com/eclipse-glsp/glsp-vscode-integration/tree/master/example/workflow). So my impl is a little bit different but I think equivalent: 

**app.ts**

```javascript
import { ContainerConfiguration } from '@eclipse-glsp/client';
import { GLSPStarter } from '@eclipse-glsp/vscode-integration-webview';
import '@eclipse-glsp/vscode-integration-webview/css/glsp-vscode.css';
import { createBPMNDiagramContainer } from '@open-bpmn/open-bpmn-glsp/lib';
import { Container } from 'inversify';

// css styles
import './css/bpmn-vscode.css';

class BPMNGLSPStarter extends GLSPStarter {
    createContainer(...containerConfiguration: ContainerConfiguration): Container {
        return createBPMNDiagramContainer(...containerConfiguration);
    }
}

export function launch(): void {
    new BPMNGLSPStarter();
}
```

**index.ts**

```javascript
import 'reflect-metadata';
import { launch } from './app';

launch();
```

and the webpack.config.js in my webview looks like this  (it is a copy from the worklfow example)

```javascript
// @ts-check
const path = require('path');

const outputPath = path.resolve(__dirname, './dist');

/**@type {import('webpack').Configuration}*/
const config = {
    target: 'web',

    entry: path.resolve(__dirname, 'src/index.ts'),
    output: {
        filename: 'webview.js',
        path: outputPath
    },
    devtool: 'eval-source-map',
    mode: 'development',

    resolve: {
        fallback: {
            fs: false,
            net: false,
        },
        alias: {
            process: 'process/browser'
        },        
        extensions: ['.ts', '.tsx', '.js']
    },
    module: {
        rules: [
            {
                test: /\.tsx?$/,
                use: ['ts-loader']
            },
            {
                test: /\.js$/,
                use: ['source-map-loader'],
                enforce: 'pre'
            },
            {
                test: /\.css$/,
                exclude: /\.useable\.css$/,
                use: ['style-loader', 'css-loader']
            }            
        ]
    },
    ignoreWarnings: [/Failed to parse source map/, /Can't resolve .* in '.*ws\/lib'/],
    performance: {
        hints: false
    }    
};

module.exports = config;
```

I am still unable to find out what should be the big difference. The only thing that differs is that I am using a Java Server. And since V2.x I did no longer find such an example using a Java server.
So maybe something is needed in addition here?

This is the current [implementation of my open-bpmn-extension.ts](https://github.com/imixs/open-bpmn-vscode-integration/blob/main/extension/src/open-bpmn-extension.ts) class.

It is really strange as I already invested hours over hours to compare the new setup with the additional webpack in the /extension/ module. But I can't figure out what I am doing wrong.

**Reply by tortmayr** [reply, marked as answer]
Ah I see.

Please ignore my comment above (`Inversify is only used in the webview (i.e. client) part of the vscode integration....`).
It looks like inversify is also needed for the extension part.
In your extension.ts class you are importing "reflect-metadata" to late: https://github.com/imixs/open-bpmn-vscode-integration/blob/4ca35122e8bc6cf4a0cae4fc898531ea0f5808cb/extension/src/open-bpmn-extension.ts#L24

It should be the first import:

```ts
import 'reflect-metadata';

import {
  GlspSocketServerLauncher,
  GlspVscodeConnector,
  SocketGlspVscodeServer,
  configureDefaultCommands
} from '@eclipse-glsp/vscode-integration/node';
import * as path from "path";
import * as process from "process";
...
```

**Reply by rsoika** [reply]
Wow!!!!!  I can't believe this! Changing the import order solved the problem.

I was not aware of this detail. I use a very new VS-Code workspace and I did recognize that the code editor formats the code and the imports. This is a useful feature for `.java` classes. But in my case VS-Code also formats `.ts` code which was fine for me. But as I now see, this it is a very dangerous feature for JavaScript.

I need to figure out which code formatting rules are applied in my VS-Code environment.

I invested more than 10 hours during the last weeks just comparing package.js and webpack.config.js files. I recognized that my imports have a different order but I was not aware that this is not correct....


@tortmayr  Thanks for your help!!!!!!
